src/app/code/community/Omise/Gateway/Controller/Base.php

```diff
@@ -1,19 +1,194 @@
 <?php
 abstract class Omise_Gateway_Controller_Base extends Mage_Core_Controller_Front_Action
 {
+    /**
+     * @var string
+     */
+    protected $title;
+
+    /**
+     * @var string
+     */
+    protected $message;
+
+    /**
+     * @var string
+     */
+    protected $transId;
+
+    /**
+     * @var Omise_Gateway_Model_Order
+     */
+    protected $order;
+
+    /**
+     * @var string
+     */
+    protected $awaitingOrderStatus;
+
+    /**
+     * @return string
+     */
+    public function getTitle()
+    {
+        return $this->title;
+    }
+
+    /**
+     * @param string $title
+     */
+    public function setTitle($title)
+    {
+        $this->title = $title;
+    }
+
+    /**
+     * @return string
+     */
+    public function getMessage()
+    {
+        return $this->message;
+    }
+
+    /**
+     * @param string $message
+     */
+    public function setMessage($message)
+    {
+        $this->message = $message;
+    }
+
+    /**
+     * @return string
+     */
+    public function getAwaitingOrderStatus()
+    {
+        return $this->awaitingOrderStatus;
+    }
+
+    /**
+     * @param string $awaitingOrderStatus
+     */
+    public function setAwaitingOrderStatus($awaitingOrderStatus)
+    {
+        $this->awaitingOrderStatus = $awaitingOrderStatus;
+    }
+
+    /**
+     * constructor
+     */
     protected function _construct()
     {
         $omise = Mage::getModel('omise_gateway/omise');
         $omise->initNecessaryConstant();
+        $this->setAwaitingOrderStatus(Mage_Sales_Model_Order::STATE_PAYMENT_REVIEW);
     }
 
     /**
-     * @return \Omise_Gateway_Model_Order
+     * @return Omise_Gateway_Model_Order
      */
     protected function _getOrder()
     {
         $id = $this->getRequest()->getParam('order_id') ? $this->getRequest()->getParam('order_id') : null;
 
         return Mage::getModel('omise_gateway/order')->getOrder($id);
     }
+
+    /**
+     * @return Mage_Core_Controller_Varien_Action|void
+     * @throws Mage_Core_Exception
+     */
+    protected function validate() {
+        if(!isset($this->order))
+            $this->order = $this->_getOrder();
+        if (! $payment = $this->order->getPayment()) {
+            Mage::getSingleton('core/session')->addError(
+                $this->__($this->getTitle().' validation failed, we cannot retrieve your payment information. 
```

Comment:
If there are strings to be inserted to translatable text, we need to avoid having those strings as part of the translation - otherwise we need to edit all the translations if (in this case) the title changes. The translation function of Magento has the facility to insert strings within a translation as follows:

`$this->__('%1 validation failed... %2...', 'string to replace %1', 'string to replace %2', ...)`

In this case though, we may also want to consider translating the title separately, so we would have:

`$this->__('%1 validation failed, we cannot retrieve your payment information. Please contact our support team to confirm the payment.', $this->__($this->getTitle())`

Although `getTitle` can probably also go (see above comment)